### DISCUSSION TITLE
Unable to prove correctness of binary_search example

### DISCUSSION THREAD

**Discussion by polwel** [flux-rs/flux#764] [General]
The documentation presents a flawed implementation of binary search. That is an awesome example!

The text mentions that an off-by-one error has been introduced. Indeed, I can tell the `while left <= right` condition should read `while left < right`. AFAICT, the implementation otherwise is identical to the one from `std`.

[However, `flux` still rejects it.](https://flux.programming.systems/?code=ZZDBUsIwEIbveYr_5LQDCOKtrb15ZoaDF8dxCmxtxlCdNMGiw7u7ySJQ7KWb_b9vs5PtxwZ2R-tc-Y5ilWXLp3BW6tOvULdY6bay-9eOKrtukkDgJiCFvp-XY_QZuEgxKbGkzhtX-E5_0xjxV-JHgT9DDlvvEHp4AE-5NdQmaT5IDdWO09mwa_VbE9rBleSr0YaELo6xXHPS9IaFCIzkzinm-QDZVUYWeWb45ZzpOkYF-ouZIsXlwugR7s7CAWT47djr2Qru0Ptbn8V_0hVJztsWi_eE2fQCPlXH55ORk7iSYII8WpuEXqoO6hc=)

What am I missing?

**Comment by nilehmann** [comment]
Hi @polwel, thanks for posting. Flux can't prove the program correct because it's missing _qualifiers_. 

To verify the program, Flux needs to infer that `size == left + right` is a loop invariant. Flux infer loops invariants by instantiating formulas from a set of "templates" called qualifiers. You can use the `scrape_quals` flag to tell Flux to try to infer those templates from the constraint. You can enable it with `#![flux::cfg(scrape_quals = true)]` ([for example](https://flux.programming.systems/?code=ZZDLbsIwEEX3_oqpKlW2gD7ozgnZdV2JRTcIoRAmxKoJ1A-atuLfO7YpEJpN7LnnzIx8ezOrte-krOo1t5Upd7j48KW2MAFnPIo5Y5vtCsweq4x5i_Ek5fQt3NnOL6FuYana0nwtLJamangA4C4QuXoeF0PoJNBBwKiAKVqvXe6t-sYhxF8BPwzo0-hg4x2EGk2nLvcaWy6yXqqxdpQ-9qtGrZtQDm5KPhulMdH5MU5jTppakRCBQZr5AOOsh-xLnRaZETw_Z6qOUQ7dRc8kxeVC6wE8nYUDoKanI68jK7h97299Ev9JVyQ6b1p4fefEigv4dDo-X2o5iislLCEvxvBQE-zAfgE=)). The flag is not enabled by default because it has performance impacts.

**Reply by ranjitjhala** [reply]
@nilehmann the link above is broken? any idea -- usual issues on `goto`?
